frontend/device/kindle/device.lua

```diff
@@ -15,10 +15,179 @@ require("ffi/fbink_input_h")
+                return {}, nil
+            end
+            local scan_result = ha_results:to_table()
+            ha_results:destroy()
+            ha_input:destroy()
+            lipc_handle:close()
+            if not scan_result then
+                -- e.g., to_table hit lha->ha == NULL
+                return {}, nil
+            else
+                return scan_result, nil
+            end
+        end
+        lipc_handle:close()
+        -- return a fake scan list containing only the currently connected profile :)
+        local profile = kindleGetCurrentProfile()
+        return { profile }, nil
+    else
+        return nil, _("Failed to acquire an anonymous lipc handle")
+    end
+end
+
+local function kindleScanThenGetResults()
+    local _ = require("gettext")
+    local haslipc, lipc = pcall(require, "liblipclua")
+    local lipc_handle
+    if haslipc then
+        lipc_handle = lipc.init("com.github.koreader.networkmgr")
+    end
+    if not lipc_handle then
+        return nil, _("Failed to acquire a lipc handle for NetworkMgr")
+    end
+
+    lipc_handle:set_string_property("com.lab126.wifid", "scan", "") -- trigger a scan
+
+    -- Mimic WpaClient:scanThenGetResults: block while waiting for the scan to finish.
+    -- Ideally, we'd do this via a poll/event workflow, but, eh', this is going to be good enough for now ;p.
+    -- For future reference, see `lipc-wait-event -m -s 0 -t com.lab126.wifid '*'`
+    --[[
+        -- For a connection:
+        [00:00:04.675699] cmStateChange "PENDING"
+        [00:00:04.677402] scanning
+        [00:00:05.488043] scanComplete
+        [00:00:05.973188] cmConnected
+        [00:00:05.977862] cmStateChange "CONNECTED"
+        [00:00:05.980698] signalStrength "1/5"
+        [00:00:06.417549] cmConnected
+
+        -- And a disconnection:
+        [00:01:34.094652] cmDisconnected
+        [00:01:34.096088] cmStateChange "NA"
+        [00:01:34.219802] signalStrength "0/5"
+        [00:01:34.221802] cmStateChange "READY"
+        [00:01:35.656375] cmIntfNotAvailable
+        [00:01:35.658710] cmStateChange "NA"
+    --]]
+    local done_scanning = false
+    local wait_cnt = 80 -- 20s in chunks on 250ms
+    while wait_cnt > 0 do
+        local scan_state = lipc_handle:get_string_property("com.lab126.wifid", "scanState")
+
+        if scan_state == "idle" then
+            done_scanning = true
+            logger.dbg("kindleScanThenGetResults: Wi-Fi scan took", (80 - wait_cnt) * 0.25, "seconds")
+            break
+        end
+
+        -- Whether it's still "scanning" or in whatever other state we don't know about,
+        -- try again until it says it's done.
+        wait_cnt = wait_cnt - 1
+        C.usleep(250 * 1000)
+    end
+    lipc_handle:close()
+
+    if done_scanning then
+        return kindleGetScanList()
+    else
+        logger.warn("kindleScanThenGetResults: Timed-out scanning for Wi-Fi networks")
+        return nil, _("Timed-out scanning for Wi-Fi networks")
```

Comment:
I'd say one of these.
```suggestion
        return nil, _("Timed out while scanning for Wi-Fi networks")
```

```suggestion
        return nil, _("Scanning for Wi-Fi networks timed out")
```